x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateUniqueIndexNameStep.java

```diff
@@ -0,0 +1,154 @@
+        ActionRequestValidationException validationException = validateGeneratedIndexName(generatedIndexName, clusterState);
+        if (validationException != null) {
+            logger.warn("unable to generate a valid index name as part of policy [{}] for index [{}] due to [{}]",
+                policy, index.getName(), validationException.getMessage());
+            throw validationException;
+        }
+        lifecycleStateSetter.apply(generatedIndexName, newCustomData);
+
+        IndexMetadata.Builder indexMetadataBuilder = IndexMetadata.builder(indexMetadata);
+        indexMetadataBuilder.putCustom(ILM_CUSTOM_METADATA_KEY, newCustomData.build().asMap());
+        newClusterStateBuilder.metadata(Metadata.builder(clusterState.getMetadata()).put(indexMetadataBuilder));
+        return newClusterStateBuilder.build();
+    }
+
+    @Nullable
+    static ActionRequestValidationException validateGeneratedIndexName(String generatedIndexName, ClusterState state) {
+        ActionRequestValidationException err = new ActionRequestValidationException();
+        try {
+            MetadataCreateIndexService.validateIndexOrAliasName(generatedIndexName, InvalidIndexNameException::new);
+        } catch (InvalidIndexNameException e) {
+            err.addValidationError(e.getMessage());
+        }
+        if (state.routingTable().hasIndex(generatedIndexName) || state.metadata().hasIndex(generatedIndexName)) {
+            err.addValidationError("the index name we generated [" + generatedIndexName + "] already exists");
+        }
+        if (state.metadata().hasAlias(generatedIndexName)) {
+            err.addValidationError("the index name we generated [" + generatedIndexName + "] already exists as alias");
+        }
+
+        if (err.validationErrors().size() > 0) {
+            return err;
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        GenerateUniqueIndexNameStep that = (GenerateUniqueIndexNameStep) o;
+        return super.equals(o) && Objects.equals(prefix, that.prefix);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), prefix);
+    }
+
+    /**
+     * This generates a valid unique index name by using the provided prefix, appended with a generated UUID, and the index name.
+     */
+    static String generateValidIndexName(String prefix, String indexName) {
+        String randomUUID = generateValidIndexSuffix(() -> UUIDs.randomBase64UUID().toLowerCase(Locale.ROOT));
+        randomUUID = randomUUID.substring(0, Math.min(randomUUID.length(), MAX_GENERATED_UUID_LENGTH));
+        return prefix + randomUUID + "-" + indexName;
+    }
+
+    static String generateValidIndexSuffix(Supplier<String> randomGenerator) {
```

Comment:
I pondered that too, but I still think the suffix is still a better name as it might not be valid if it'd be used as a prefix (ie. we don't make sure the generated UUID doesn't *start with* `_` or `-`)

Reply:
Okay, that's fine with me :)